### PR TITLE
FTR-12 Notifications

### PR DESCRIPTION
**Pull request by dpetek21**
For the consideration of the midnight society, the Notification feature is done on the Kotlin side of things. Merging will add the capability of displaying Firebase Messaging notifications to the ServiceSync app, as well as automatic saving of the user's device's Firebase messaging ID upon Login.

### CODE REVIEW

File: app/src/main/java/hr/foi/air/servicesync/backend/FirestoreUserDetails.kt [hunk 1]
```diff
@@ -74,4 +76,57 @@ class FirestoreUserDetails {
                 onResult(false)
             }
     }
+    //FOR MESSAGING
+    fun saveFCMToken(token: String,onResult: (Boolean) -> Unit) {
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/backend/FirestoreUserDetails.kt:80]: Well made backend code that can be used properly considering OOP.

File: app/src/main/java/hr/foi/air/servicesync/backend/firebaseAuthentication.kt [hunk 2]
```diff
@@ -12,6 +12,8 @@ class firebaseAuthentication {
             .addOnCompleteListener { task ->
                 if (task.isSuccessful) {
                     val user = auth.currentUser
+                    val firesoreUserDetails : FirestoreUserDetails = FirestoreUserDetails()
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/backend/firebaseAuthentication.kt:15]: Well done decomposing usage of a such function to various layers of the application.

File: app/src/main/java/hr/foi/air/servicesync/backend/firebaseMessagingHandler.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import android.os.Build
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import android.util.Log
+import androidx.core.app.ActivityCompat
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import hr.foi.air.servicesync.R
+
+class firebaseMessagingHandler: FirebaseMessagingService(){
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/backend/firebaseMessagingHandler.kt:15]: This is how we do it!

File: app/src/main/java/hr/foi/air/servicesync/data/UserSession.kt [hunk 4]
```diff
@@ -11,6 +12,9 @@ object UserSession {
     }
 
     fun logout() {
-        FirebaseAuth.getInstance().signOut()
+        val firesoreUserDetails : FirestoreUserDetails = FirestoreUserDetails()
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/data/UserSession.kt:15]: Revamped logout functionality according to changes made for notifications, nice!

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt [hunk 5]
```diff
@@ -118,6 +123,41 @@ fun LoginScreen(
 
                 Button(
                     onClick =  {
+                        /*
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt:126]: Consider removing commented code if it isn't necessary.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt [hunk 6]
```diff
@@ -118,6 +123,41 @@ fun LoginScreen(
...
+                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                            ActivityCompat.requestPermissions(
+                                this,
+                                arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
+                                REQUEST_CODE
+                            )
+                        }*/
+                        val activity = context as? android.app.Activity
+                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                            if (ActivityCompat.checkSelfPermission(
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt:136]: Interesting, selfpermission, interesting indeed.

**Review by ibeusan20 — APPROVED** [review]
Well made, high quality code. You can literally see the effort! Very nice 👍

File: app/src/main/java/hr/foi/air/servicesync/backend/FirestoreUserDetails.kt [hunk 7]
```diff
@@ -74,4 +76,57 @@ class FirestoreUserDetails {
...
+                            onResult(true)
+                        }
+                        onResult(false)
+                    }
+                } else {
+                    Log.e("FCM", "Failed to retrieve FCM token", task.exception)
+                    onResult(false)
+                }
+            }
+    }
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/backend/FirestoreUserDetails.kt:131]: Very good, token is saved on login, updated and deleted on logout if the user logs in on another device.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt (hunk 5 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt:126]: Agreed.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt [hunk 8]
```diff
@@ -118,6 +123,41 @@ fun LoginScreen(
...
+                        }
+                        /*FirebaseMessaging.getInstance().token
+                            .addOnCompleteListener { task ->
+                                if (task.isSuccessful) {
+                                    val token = task.result
+                                    Log.d("FCM", "FCM Token: $token")
+                                } else {
+                                    Log.e("FCM", "Failed to retrieve FCM token", task.exception)
+                                }
+                            }*/
```
Comment (mpuskadij) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/LoginScreen.kt:160]: Remove this if unnecessary.

**Review by mpuskadij — APPROVED** [review]
Very well done. This completes the app side of notifications, a server will need to be developed to perform checking if 24 (example) hours remain until the appointment and then send a notification to the firebase

**Pull request merged**